### DISCUSSION TITLE
UPF powerswitches and PDN alignment

### DISCUSSION THREAD

**Discussion by kareefardi** [The-OpenROAD-Project/OpenROAD#5762] [Q&A]
Using the abstract view of a powerswitch inside the testcases of PDN, the powerswitch is placed on the wrong row leading to wrong switched power/gnd connection to the standard cells in the design. I have tried `power_switch_upf_daisy` testcase and another testcase on my own. Here are the results: 
![image](https://github.com/user-attachments/assets/bac47262-f92e-45cb-9286-3c6232458f5e)
Highlighted in red is the power pin of the powerswitch and in green is the power rail. The power rails and pins of the switch don't align. I am not sure if this is a bug or if this is somehow controllable in a way that I do not know about so I opened a discussion.

**Comment by maliberty** [comment]
Is it possible to make an issue with a test case?

@gadfort any thoughts?

**Reply by gadfort** [reply]
@maliberty since the power switches are inserted on the double height grid (assuming there isn't a bug in code), it would imply that the double height grid doesn't align with the single height grid. (ie. they are off by half a row).
A test case would be required to further debug this.

**Reply by gadfort** [reply]
@maliberty I actually think the cell is wrong, there is no way to insert this power switch into the rows without causing that issues. It has power on top and bottom, but the rows are aligned to the grounds (the switch cell is wrong and should be updated. I can do that later)

**Reply by gadfort** [reply]
https://github.com/The-OpenROAD-Project/OpenROAD/pull/5766

**Reply by kareefardi** [reply]
> but the rows are aligned to the grounds

Is there a way to make them aligned to the power?. If we have single row design. It can start with ground or power if the cells get flipped? Just wondering here

**Reply by gadfort** [reply]
Since PDN only uses the rows provided, it would be up to the row generation to provide this alignment

**Comment by maliberty** [comment]
I'm not sure what I'm looking at in the image.  A single quad-height cell?  What is its orientation?  How are the rows defined?  I find I can't provide any help without a concrete test case to look at.

**Reply by kareefardi** [reply]
> I'm not sure what I'm looking at in the image. A single quad-height cell? What is its orientation? How are the rows defined? I find I can't provide any help without a concrete test case to look at.

As for the image. The power switch is represented as a double height cell. The rails of the powers switch are represented as POWER-GROUND-POWER. They weren't aligned with the std cells rails (single height). The PR above switched the power switch to GROUND-POWER-GROUND.

Everything was generated using the script `src/pdn/test/power_switch_upf_daisy.tcl` with the modification that I highlighted below.

My inquiry is whether this is a hard requirement. I don't see anything wrong with having the std cell rails start with POWER in order to align with the power switches. However, I am not fully familiar with the problem (powerswitches, rails, etc. ) so I am phrasing this as an inquiry/discussion.

**Comment by kareefardi** [comment]
Before the change above, I was running a testcase in OpenROAD `src/pdn/test/power_switch_upf_daisy.tcl` and modified it by adding 
```
global_placement
detailed_placement
```
and inspecting the results
